### ISSUE TITLE
Forward decls shouldn't complete anything. [phab:D1623]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2047]
_Created by **salguarnieri** at 2017-05-06 00:14:02_: 

When clang first encounters a forward decl (class FooClass),
hasDefinition returns false, which allows you to differentiate between
class decls and interface decls. The definition should always be the
implementation block. After clang finds a interface, the class
forward declaration object's hasDefinition will return true, so it can
no longer be used to differentiate between forward declarations and
proper declarations.

To get around this, we differentiate between forward declarations and
the full declaration by checking to see if the decl is the canonical
decl. The interface decl should be the canonical decl. Only the
canonical decl "complete"s other decls.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2017-05-06 00:15:20_:

rebase

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2017-05-06 00:17:42_:

  - remove old comments

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2017-05-06 00:52:47_:

  - add another case to test

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2017-05-06 03:14:49_:

  - Add some more comments

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-05-08 20:41:10_:

FwdIface is a singleton here; what is it meant to unify with?

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2017-05-08 20:45:16_:

I could make it _ since the main goal is to just say that something completes FwdDecl. But if there is a way to associate it with the anchor text from FwdClass.h I'd like to do that.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-05-08 21:23:51_:

You could use

```
//- FwdIface=vname(_,_,_,"kythe/cxx/indexer/cxx/testdata/objc/FwdClass.h",_) completes FwdDecl
//- FwdIface.node/kind anchor
```

(or convert the test to a bundle test, or change the rules to permit multiple source files or reading from file facts, but this seems like the simplest option).

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2017-05-08 21:28:41_:

  - restrict what FwdIface can bind to

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2017-05-08 21:29:13_:

Thanks. Yeah, I didn't think it was worth going all the way to a bundle test.

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2017-05-08 21:40:27_

**Comment by kythe** [comment]
_**salguarnieri** closed via commit ca57c386e6a26446e876171e05467b631995aef3 at 2017-05-08 21:42:03_
